<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Benchmarks: discretize() • FSelectorRcpp</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">FSelectorRcpp</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/index.html">Articles</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><link href="benchmarks_discretize_files/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Benchmarks: discretize()</h1>
            <h3 class="subtitle">
<small> <a href="http://r-addict.com/About.html">Marcin Kosiński</a>   <a href="https://stackoverflow.com/users/3857701"><i class="fa fa-stack-overflow"></i></a>  <a href="https://r-addict.com"><i class="fa fa-comment"></i></a>  <a href="https://github.com/MarcinKosinski"><i class="fa fa-github"></i></a>  <a href="mailto:m.p.kosinski@gmail.com"><i class="fa fa-envelope-o"></i></a> <br><a href="https://www.linkedin.com/in/krzysztof-s%C5%82omczy%C5%84ski-20a364134">Krzysztof Słomczyński</a>   <a href="https://github.com/krzyslom"><i class="fa fa-github"></i></a>  <a href="mailto:krzysztofslomczynski@gmail.com"><i class="fa fa-envelope-o"></i></a></small><br>
</h3>
            
          </div>

    
    
<div class="contents">
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(microbenchmark)
<span class="kw">library</span>(FSelectorRcpp)
<span class="kw">library</span>(RWeka)</code></pre></div>
<div id="about-discretization" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#about-discretization" class="anchor"> </a></body></html>About discretization</h1>
<p>In the statistical modelling, the <strong>discretization</strong> is the process of transferring continuous explanatory variables into discrete counterparts. Problems caused by categorization of continuous variables are known and widely spread <span class="citation">(Harrell 2015)</span>, but in some cases there appear an algorithmic requirement for the discretization. Moreover, there exist few algorithms, like <strong>Decision Trees</strong> <span class="citation">(Salzberg 1994)</span>, where continuous attributes are discretized during the learning process. Other reason for variable discretization include increasing the speed of induction algorithms <span class="citation">(Catlett 1991)</span>.</p>
<p>Even though many categorization algorithms have been developed <span class="citation">(Holte 1993, <span class="citation">Chan, Batur, and Srinivasan (1991)</span>)</span>, in this article we focus on a recursive entropy minimization heuristic for categorization coupled with a <strong>Minimum Description Length (MDL)</strong> critetion <span class="citation">(Rissanen 1986)</span> that controls the number of intervals produced over the continuous space. This is motivated by the original usage of this algorithm in <strong>FSelector</strong> package, which is our baseline.</p>
<p><span class="citation">(Dougherty, Kohavi, and Sahami 1995)</span> showed better performance of classification for discretized feature set on real-world datasets and states that the described method was found promising by the authors not only for the local discretization but also for global discretization <span class="citation">(Ting 1994)</span>.</p>
</div>
<div id="comparisons-on-rtcga-data" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#comparisons-on-rtcga-data" class="anchor"> </a></body></html>Comparisons on RTCGA data</h1>
<div id="about-the-cancer-genome-atlas" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#about-the-cancer-genome-atlas" class="anchor"> </a></body></html>About The Cancer Genome Atlas</h2>
<blockquote>
<p>The Cancer Genome Atlas (TCGA) is a comprehensive and coordinated effort to accelerate our understanding of the molecular basis of cancer through the application of genome analysis technologies, including large-scale genome sequencing - <a href="http://cancergenome.nih.gov/" class="uri">http://cancergenome.nih.gov/</a>.</p>
</blockquote>
<p><a href="https://github.com/orgs/RTCGA/people">RTCGA team</a> converted selected datasets from this study into few separate packages that are hosted on Bioconductor. These R packages make selected datasets easier to access and manage. Data sets in RTCGA packages are large and cover complex relations between clinical outcomes and genetic background.</p>
<p>To use RTCGA install package with instructions from it’s <a href="https://www.bioconductor.org/packages/RTCGA/">Bioconductor home page</a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## try http:// if https:// URLs are not supported
<span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"RTCGA"</span>)</code></pre></div>
</div>
<div id="dependent-variable" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#dependent-variable" class="anchor"> </a></body></html>Dependent variable</h2>
<p>Specifies whether a sample came from tumour or healthy tissue. It is denoted by 14th sign in <code>bcr_patiend_barcode</code> variable in <code>BRCA.rnaseq</code> dataset that containt RNA-Seq genes’ expressions for patients suffering from Breast Cancer.</p>
</div>
<div id="independent-variables" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#independent-variables" class="anchor"> </a></body></html>Independent variables</h2>
<p>It is mandatory to get rid of special signs from their names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(RTCGA.rnaseq)
BRCA.rnaseq &lt;-<span class="st"> </span>RTCGA.rnaseq::BRCA.rnaseq
BRCA.rnaseq$bcr_patient_barcode &lt;-<span class="st"> </span>
<span class="st">   </span><span class="kw">factor</span>(<span class="kw">substr</span>(BRCA.rnaseq$bcr_patient_barcode, <span class="dv">14</span>, <span class="dv">14</span>))
<span class="kw">names</span>(BRCA.rnaseq) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">"?"</span>, <span class="dt">replacement =</span> <span class="st">"q"</span>,
                           <span class="dt">x =</span> <span class="kw">names</span>(BRCA.rnaseq), <span class="dt">fixed =</span> <span class="ot">TRUE</span>)
<span class="kw">names</span>(BRCA.rnaseq) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">"[[:punct:]]"</span>, <span class="dt">replacement =</span> <span class="st">"_"</span>,
                           <span class="dt">x =</span> <span class="kw">names</span>(BRCA.rnaseq))</code></pre></div>
</div>
<div id="more" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#more" class="anchor"> </a></body></html>More</h2>
<ul>
<li><a href="http://r-addict.com/2016/05/04/RTCGA-Quick-Guide.html">RTCGA Quick Guide</a></li>
<li><a href="http://rtcga.github.io/RTCGA/">RTCGA Website</a></li>
</ul>
</div>
</div>
<div id="discretization-benchmarks" class="section level1">
<h1 class="hasAnchor">
<html><body><a href="#discretization-benchmarks" class="anchor"> </a></body></html>Discretization Benchmarks</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
<span class="kw">library</span>(pbapply)
<span class="kw">library</span>(tibble)
names_by &lt;-<span class="st"> </span>function(nameBy, vecBy) {
  <span class="kw">c</span>(<span class="kw">paste0</span>(nameBy,
           <span class="kw">sapply</span>(<span class="kw">max</span>(<span class="kw">nchar</span>(vecBy)) -<span class="st"> </span><span class="kw">nchar</span>(vecBy),
                  function(zeros) {
                    <span class="kw">paste0</span>(<span class="kw">rep</span>(<span class="dv">0</span>, zeros), <span class="dt">collapse =</span> <span class="st">""</span>)
                  }), 
                  vecBy))
}

get_times &lt;-<span class="st"> </span>function(fun, vecRows, vecCols, nTimes) {
  forRows &lt;-<span class="st"> </span><span class="kw">pblapply</span>(vecRows, function(nRows) {
    forCols &lt;-<span class="st"> </span><span class="kw">pblapply</span>(vecCols +<span class="st"> </span><span class="dv">1</span>, function(nCols) {
      <span class="kw">microbenchmark</span>(<span class="kw">fun</span>(bcr_patient_barcode ~<span class="st"> </span>.,
                         BRCA.rnaseq[<span class="dv">1</span>:nRows, <span class="dv">1</span>:nCols]),
                     <span class="dt">times =</span> nTimes)
    })
    colsNames &lt;-<span class="st"> </span><span class="kw">names_by</span>(<span class="dt">nameBy =</span> <span class="st">"columns_"</span>, <span class="dt">vecBy =</span> vecCols)
    <span class="kw">setNames</span>(forCols, colsNames)
  })
  rowsNames &lt;-<span class="st"> </span><span class="kw">names_by</span>(<span class="dt">nameBy =</span> <span class="st">"rows_"</span>, <span class="dt">vecBy =</span> vecRows)
  <span class="kw">setNames</span>(forRows, rowsNames)
}

make_df &lt;-<span class="st"> </span>function(bm_data, package) {
  <span class="kw">data_frame</span>(
    <span class="dt">rows =</span> <span class="kw">rep</span>(
      <span class="dt">x =</span> <span class="kw">as.integer</span>(
        <span class="kw">gsub</span>(
          <span class="dt">pattern =</span> <span class="st">"rows_0*"</span>,
             <span class="dt">replacement =</span> <span class="st">""</span>,
             <span class="dt">x =</span> <span class="kw">names</span>(bm_data)
        )
      ),
      <span class="dt">each =</span>  <span class="kw">unique</span>(<span class="kw">sapply</span>(bm_data, length))
    ),
    <span class="dt">columns =</span> <span class="kw">as.integer</span>(
      <span class="kw">gsub</span>(
        <span class="dt">pattern =</span> <span class="st">"columns_0*"</span>,
        <span class="dt">replacement =</span> <span class="st">""</span>,
        <span class="dt">x =</span> <span class="kw">as.vector</span>(<span class="kw">sapply</span>(bm_data, names))
      )
    ),
    <span class="dt">time =</span> <span class="kw">as.vector</span>(
      <span class="kw">sapply</span>(
        bm_data,
        function(setOfCols) {
          <span class="kw">sapply</span>(
            setOfCols,
            function(set) {
              <span class="kw">median</span>(set$time) /<span class="st"> </span><span class="dv">10</span> ^<span class="st"> </span><span class="dv">9</span>
            }
          )
        }
      )
    ),
    <span class="dt">package =</span> package
  )
}

bm_fsr_discretize &lt;-<span class="st"> </span><span class="kw">get_times</span>(<span class="dt">fun =</span> discretize,
                               <span class="dt">vecRows =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">200</span>, <span class="dt">to =</span> <span class="dv">1000</span>, <span class="dt">by =</span> <span class="dv">200</span>),
                               <span class="dt">vecCols =</span> <span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">500</span>, <span class="dv">1000</span>),
                               <span class="dt">nTimes =</span> 1L)</code></pre></div>
<pre><code>
   |+++++++++++++                                     | 25% ~00s          
   |+++++++++++++++++++++++++                         | 50% ~00s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~00s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 01s

   |++++++++++                                        | 20% ~04s          
   |+++++++++++++                                     | 25% ~00s          
   |+++++++++++++++++++++++++                         | 50% ~00s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~00s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 01s

   |++++++++++++++++++++                              | 40% ~03s          
   |+++++++++++++                                     | 25% ~00s          
   |+++++++++++++++++++++++++                         | 50% ~00s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~00s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 01s

   |++++++++++++++++++++++++++++++                    | 60% ~02s          
   |+++++++++++++                                     | 25% ~00s          
   |+++++++++++++++++++++++++                         | 50% ~00s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~00s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 02s

   |++++++++++++++++++++++++++++++++++++++++         | 80% ~01s          
   |+++++++++++++                                     | 25% ~00s          
   |+++++++++++++++++++++++++                         | 50% ~00s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~00s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 02s

   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 07s</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bm_rwk_discretize &lt;-<span class="st"> </span><span class="kw">get_times</span>(<span class="dt">fun =</span> Discretize,
                               <span class="dt">vecRows =</span> <span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">200</span>, <span class="dt">to =</span> <span class="dv">1000</span>, <span class="dt">by =</span> <span class="dv">200</span>),
                               <span class="dt">vecCols =</span> <span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">100</span>, <span class="dv">500</span>, <span class="dv">1000</span>),
                               <span class="dt">nTimes =</span> 1L)</code></pre></div>
<pre><code>
   |+++++++++++++                                     | 25% ~02s          
   |+++++++++++++++++++++++++                         | 50% ~01s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~01s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 04s

   |++++++++++                                        | 20% ~15s          
   |+++++++++++++                                     | 25% ~01s          
   |+++++++++++++++++++++++++                         | 50% ~00s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~00s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 03s

   |++++++++++++++++++++                              | 40% ~11s          
   |+++++++++++++                                     | 25% ~01s          
   |+++++++++++++++++++++++++                         | 50% ~00s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~01s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 04s

   |++++++++++++++++++++++++++++++                    | 60% ~07s          
   |+++++++++++++                                     | 25% ~01s          
   |+++++++++++++++++++++++++                         | 50% ~01s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~01s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 05s

   |++++++++++++++++++++++++++++++++++++++++         | 80% ~04s          
   |+++++++++++++                                     | 25% ~01s          
   |+++++++++++++++++++++++++                         | 50% ~01s          
   |++++++++++++++++++++++++++++++++++++++            | 75% ~01s          
   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 05s

   |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 21s</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">df_fsr &lt;-<span class="st"> </span><span class="kw">make_df</span>(<span class="dt">bm_data =</span> bm_fsr_discretize, <span class="dt">package =</span> <span class="st">"FSelectorRcpp"</span>)
df_rwk &lt;-<span class="st"> </span><span class="kw">make_df</span>(<span class="dt">bm_data =</span> bm_rwk_discretize, <span class="dt">package =</span> <span class="st">"RWeka"</span>)
bm_df &lt;-<span class="st"> </span><span class="kw">rbind</span>(df_fsr, df_rwk)

make_plot &lt;-<span class="st"> </span>function(y) {
  yUnit &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">"time"</span>,
                <span class="dt">replacement =</span> <span class="st">"s"</span>,
                <span class="dt">x =</span> y)
  
  <span class="kw">ggplot</span>(<span class="dt">data =</span> bm_df, <span class="kw">aes_string</span>(<span class="dt">x =</span> <span class="st">"rows"</span>, <span class="dt">y =</span> y, <span class="dt">color =</span> <span class="st">"package"</span>)) +
<span class="st">  </span><span class="kw">facet_grid</span>(columns ~<span class="st"> </span>., <span class="dt">labeller =</span> label_both) +
<span class="st">  </span><span class="kw">geom_point</span>() +
<span class="st">  </span><span class="kw">xlab</span>(<span class="st">"Number of rows [-]"</span>) +
<span class="st">  </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">"Elapsed time ["</span>, yUnit, <span class="st">"]"</span>)) +
<span class="st">  </span><span class="kw">ggtitle</span>(<span class="dt">label =</span> <span class="st">"Microbenchmark: discretization function comparison"</span>,
          <span class="dt">subtitle =</span> <span class="st">"for different sizes of data sets"</span>) +
<span class="st">  </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),
        <span class="dt">plot.subtitle =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),
        <span class="dt">legend.position =</span> <span class="st">"top"</span>)
}

bm_plot &lt;-<span class="st"> </span><span class="kw">make_plot</span>(<span class="st">"time"</span>)
bm_plot_log &lt;-<span class="st"> </span><span class="kw">make_plot</span>(<span class="st">"log(time)"</span>)</code></pre></div>
<div id="plots" class="section level2">
<h2 class="hasAnchor">
<html><body><a href="#plots" class="anchor"> </a></body></html>Plots</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bm_plot</code></pre></div>
<p><img src="benchmarks_discretize_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bm_plot_log</code></pre></div>
<p><img src="benchmarks_discretize_files/figure-html/unnamed-chunk-5-2.png" width="768"></p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<html><body><a href="#references" class="anchor"> </a></body></html>References</h1>
<div id="refs" class="references">
<div id="ref-Catlett1991">
<p>Catlett, J. 1991. “On Changing Continuous Attributes into Ordered Discrete Attributes.” In <em>Machine Learning — Ewsl-91: European Working Session on Learning Porto, Portugal, March 6–8, 1991 Proceedings</em>, edited by Yves Kodratoff, 164–78. Berlin, Heidelberg: Springer Berlin Heidelberg. doi:<a href="https://doi.org/10.1007/BFb0017012">10.1007/BFb0017012</a>.</p>
</div>
<div id="ref-Chan169942">
<p>Chan, C. C., C. Batur, and A. Srinivasan. 1991. “Determination of Quantization Intervals in Rule Based Model for Dynamic Systems.” In <em>Systems, Man, and Cybernetics, 1991. ’Decision Aiding for Complex Systems, Conference Proceedings., 1991 Ieee International Conference on</em>, 1719–23 vol.3. doi:<a href="https://doi.org/10.1109/ICSMC.1991.169942">10.1109/ICSMC.1991.169942</a>.</p>
</div>
<div id="ref-Dougherty95supervisedand">
<p>Dougherty, James, Ron Kohavi, and Mehran Sahami. 1995. “Supervised and Unsupervised Discretization of Continuous Features.” In <em>MACHINE Learning: PROCEEDINGS of the Twelfth International Conference</em>, 194–202. Morgan Kaufmann.</p>
</div>
<div id="ref-HarrellDISC">
<p>Harrell, Frank. 2015. “General Handouts and Reference Material on Biostatistical Modeling - Problems Caused by Categorizing Continuous Variables.” <em>Vanderbilt Biostatistics Wiki</em>. <a href="http://biostat.mc.vanderbilt.edu/wiki/Main/CatContinuous" class="uri">http://biostat.mc.vanderbilt.edu/wiki/Main/CatContinuous</a>.</p>
</div>
<div id="ref-Holte1993">
<p>Holte, Robert C. 1993. “Very Simple Classification Rules Perform Well on Most Commonly Used Datasets.” <em>Machine Learning</em> 11 (1): 63–90. doi:<a href="https://doi.org/10.1023/A:1022631118932">10.1023/A:1022631118932</a>.</p>
</div>
<div id="ref-rissanen1986">
<p>Rissanen, Jorma. 1986. “Stochastic Complexity and Modeling.” <em>Ann. Statist.</em> 14 (3). The Institute of Mathematical Statistics: 1080–1100. doi:<a href="https://doi.org/10.1214/aos/1176350051">10.1214/aos/1176350051</a>.</p>
</div>
<div id="ref-Salzberg1994">
<p>Salzberg, Steven L. 1994. “C4.5: Programs for Machine Learning by J. Ross Quinlan. Morgan Kaufmann Publishers, Inc., 1993.” <em>Machine Learning</em> 16 (3): 235–40. doi:<a href="https://doi.org/10.1007/BF00993309">10.1007/BF00993309</a>.</p>
</div>
<div id="ref-Ting94discretizationof">
<p>Ting, Kai Ming. 1994. “Discretization of Continuous-Valued Attributes and Instance-Based Learning.”</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#about-discretization">About discretization</a></li>
      <li>
<a href="#comparisons-on-rtcga-data">Comparisons on RTCGA data</a><ul class="nav nav-pills nav-stacked">
<li><a href="#about-the-cancer-genome-atlas">About The Cancer Genome Atlas</a></li>
      <li><a href="#dependent-variable">Dependent variable</a></li>
      <li><a href="#independent-variables">Independent variables</a></li>
      <li><a href="#more">More</a></li>
      </ul>
</li>
      <li>
<a href="#discretization-benchmarks">Discretization Benchmarks</a><ul class="nav nav-pills nav-stacked">
<li><a href="#plots">Plots</a></li>
      </ul>
</li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Zygmunt Zawadzki, Marcin Kosinski.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
